processing/src/main/java/org/apache/carbondata/processing/newflow/steps/input/InputProcessorStepImpl.java

```diff
@@ -0,0 +1,171 @@
+    }
+    return numberOfCores;
+  }
+
+  private int getBatchSize() {
+    int batchSize;
+    try {
+      batchSize = Integer.parseInt(configuration
+          .getDataLoadProperty(DataLoadProcessorConstants.DATA_LOAD_BATCH_SIZE,
+              DataLoadProcessorConstants.DATA_LOAD_BATCH_SIZE_DEFAULT).toString());
+    } catch (NumberFormatException exc) {
+      batchSize = Integer.parseInt(DataLoadProcessorConstants.DATA_LOAD_BATCH_SIZE_DEFAULT);
+    }
+    return batchSize;
+  }
+
+  @Override public Iterator<CarbonRowBatch>[] execute() {
+    int batchSize = getBatchSize();
+    List<Iterator<Object[]>>[] readerIterators = partitionInputReaderIterators();
+    Iterator<CarbonRowBatch>[] outIterators = new Iterator[readerIterators.length];
+    for (int i = 0; i < outIterators.length; i++) {
+      outIterators[i] = new InputProcessorIterator(readerIterators[i], genericParsers, batchSize);
+    }
+    return outIterators;
+  }
+
+  private List<Iterator<Object[]>>[] partitionInputReaderIterators() {
+    int numberOfCores = getNumberOfCores();
+    if (inputIterators.size() < numberOfCores) {
+      numberOfCores = inputIterators.size();
+    }
+    List<Iterator<Object[]>>[] iterators = new List[numberOfCores];
+    for (int i = 0; i < numberOfCores; i++) {
+      iterators[i] = new ArrayList<>();
+    }
+
+    for (int i = 0; i < inputIterators.size(); i++) {
+      iterators[i % numberOfCores].add(inputIterators.get(i));
+
+    }
+    return iterators;
+  }
+
+  @Override protected CarbonRow processRow(CarbonRow row) {
+    return null;
+  }
+
+  private static class InputProcessorIterator extends CarbonIterator<CarbonRowBatch> {
+
+    private List<Iterator<Object[]>> inputIterators;
+
+    private GenericParser[] genericParsers;
+
+    private Iterator<Object[]> currentIterator;
+
+    private int counter;
+
+    private int batchSize;
+
+    public InputProcessorIterator(List<Iterator<Object[]>> inputIterators,
+        GenericParser[] genericParsers, int batchSize) {
+      this.inputIterators = inputIterators;
+      this.genericParsers = genericParsers;
+      this.batchSize = batchSize;
+      currentIterator = inputIterators.get(counter++);
+    }
+
+    @Override public boolean hasNext() {
+      return internalHasNext();
+    }
+
+    private boolean internalHasNext() {
+      boolean hasNext = currentIterator.hasNext();
+      if (!hasNext) {
+        if (counter < inputIterators.size()) {
+          currentIterator = inputIterators.get(counter++);
+        }
+        hasNext = internalHasNext();
+      }
+      return hasNext;
+    }
+
+    @Override public CarbonRowBatch next() {
+      CarbonRowBatch carbonRowBatch = new CarbonRowBatch();
+      int count = 0;
+      while (internalHasNext() && count < batchSize) {
+        Object[] row = currentIterator.next();
+        for (int i = 0; i < row.length; i++) {
+          row[i] = genericParsers[i].parse(row[i].toString());
```

Comment:
This will involve calling many virtual function call, can we make it a RowParser and RowParserFactory, then later we can add code generation for it.